### PR TITLE
[shopsys] some little adjustments in js and webpack

### LINKED ISSUE #1678: Quantity change button does not work correctly
### What is happening
Front end -> Product detail, Product list, Cart

When I hold button for adding quantity, the counting starts and cannot be stopped.

### Expected result
Please, change quantity only by click.

### LINKED ISSUE #1687: wrong usage of javascript in template
### What is happening
There is a `Shopsys.window` call in https://github.com/shopsys/shopsys/blob/master/packages/framework/src/Resources/views/Admin/Content/Category/list.html.twig#L36 - this should not be here anymore after https://github.com/shopsys/shopsys/pull/1545 was merged

### Expected result
The JS call is fixed :slightly_smiling_face:

### LINKED ISSUE #1711: JS overridding not working
### What is happening
https://docs.shopsys.com/en/latest/frontend/npm-and-webpack/#i-want-to-override-class-from-shopsysframework-common-package describes how to override a js from the framework, however, when I follow the instructions, I get an error in console:
```
Uncaught TypeError: Class constructor CategoryTreeSorting cannot be invoked without 'new'
    at new ExtendedCategoryTreeSorting (categoryTree.sorting.js:7)
    at Object.init [as callback] (categoryTree.sorting.js:33)
    at Register.registerNewContent (Register.js:71)
    at HTMLDocument.<anonymous> (registerAdmin.js:14)
    at mightThrow (jquery.js:3557)
    at process (jquery.js:3625)
```

Maybe this is an ES6 restriction? https://stackoverflow.com/questions/51860043/javascript-es6-typeerror-class-constructor-client-cannot-be-invoked-without-ne :thinking: 

### Expected result
It is possible to override a JS from the framework following the docs.

### PR DESCRIPTION
**Pull request by pesektomas**
| Q             | A
| ------------- | ---
|Description, reason for the PR| I accumulated points from implementation of webpack and javascript and in this PR I solved a few from it
|New feature| No 
|[BC breaks](https://docs.shopsys.com/en/latest/contributing/backward-compatibility-promise/)| No 
|Fixes issues| closes #1711, closes #1687, closes #1678, closes #1721, closes #1710
|Have you read and signed our [License Agreement for contributions](https://www.shopsys.com/license-agreement)?| Yes

### CODE REVIEW

File: project-base/src/Controller/Front/NewsletterController.php [hunk 1]
```diff
@@ -53,12 +64,43 @@ public function subscribeEmailAction(Request $request)
...
+        if ($form->isSubmitted()) {
+            $this->formTimeProvider->generateFormTime($form->getName());
+
+            if ($form->isValid()) {
+                $email = $form->getData()['email'];
+
+                try {
+                    $this->newsletterFacade->addSubscribedEmail($email, $this->domain->getId());
+                    return $this->json(['success' => true]);
+                } catch (EmailWasAlreadySubscribedException $exception) {
```
Comment (grossmannmartin) [inline project-base/src/Controller/Front/NewsletterController.php:76]: If you decide to keep the EmailAlreadySubscribed functionality, what about creating a Validator for that (similar to UniqueEmailValidator).

Because currently, when I reuse the `SubscriptionFormType` and/or `addSubscribedEmail` I can easily forget to catch this exception and application end with error 500. (plus it could make this controller method more clear)

File: project-base/src/Controller/Front/NewsletterController.php [hunk 2]
```diff
@@ -79,6 +121,7 @@ private function createSubscriptionForm(): Form
         /** @var \Symfony\Component\Form\Form $form */
         $form = $this->createForm(SubscriptionFormType::class, null, [
             'action' => $this->generateUrl('front_newsletter_send'),
+            'compound' => true,
```
Comment (grossmannmartin) [inline project-base/src/Controller/Front/NewsletterController.php:124]: Can you please explain to me this configuration? Why is it needed?

File: project-base/webpack.config.js [hunk 3]
```diff
@@ -7,6 +7,7 @@ const yaml = require('js-yaml');
 const fs = require('fs');
 const path = require('path');
 const StylelintPlugin = require('stylelint-webpack-plugin');
+const sources = require('./assets/js/bin/helpers/sources');
```
Comment (grossmannmartin) [inline project-base/webpack.config.js:10]: This constant seems unused.

File: packages/framework/assets/js/admin/registerAdmin.js [hunk 4]
```diff
@@ -8,9 +8,13 @@ import './validation/form';
 
 import Register from '../common/utils/Register';
 
-export default function registerAdmin () {
+export default function registerAdmin (afterRegistrationCallback = null) {
```
Comment (grossmannmartin) [inline packages/framework/assets/js/admin/registerAdmin.js:11]: Was this change necessary?

File: project-base/src/Controller/Front/NewsletterController.php [hunk 5]
```diff
@@ -53,12 +64,43 @@ public function subscribeEmailAction(Request $request)
         $form = $this->createSubscriptionForm();
         $form->handleRequest($request);
 
-        if ($form->isSubmitted() && $form->isValid()) {
-            $email = $form->getData()['email'];
-            $this->newsletterFacade->addSubscribedEmail($email, $this->domain->getId());
+        if ($form->isSubmitted()) {
+            $this->formTimeProvider->generateFormTime($form->getName());
```
Comment (grossmannmartin) [inline project-base/src/Controller/Front/NewsletterController.php:68]: is it necessary to generate formtime manually? I thought it's handled by listener on form pre-submit event

File: project-base/src/Controller/Front/NewsletterController.php [hunk 6]
```diff
@@ -53,12 +64,43 @@ public function subscribeEmailAction(Request $request)
...
+            if ($form->isValid()) {
+                $email = $form->getData()['email'];
+
+                try {
+                    $this->newsletterFacade->addSubscribedEmail($email, $this->domain->getId());
+                    return $this->json(['success' => true]);
+                } catch (EmailWasAlreadySubscribedException $exception) {
+                    return $this->json([
+                        'success' => false,
+                        'errors' => [t('Email was already subscribed')],
```
Comment (grossmannmartin) [inline project-base/src/Controller/Front/NewsletterController.php:79]: can you dump translations for this PR?

File: packages/framework/assets/js/admin/registerAdmin.js (hunk 4 above)
Comment (pesektomas) [reply packages/framework/assets/js/admin/registerAdmin.js:11]: It wasn't, but it occurred to me that it could be used by somebody because we have the option to influence admin only with registerCallbacks and priorities now.

File: packages/framework/build.xml [hunk 7]
```diff
@@ -15,9 +15,12 @@
     <property name="path.composer.executable" value="composer"/>
     <property name="path.config" value="${path.root}/config"/>
     <property name="path.env.test" value="${path.root}/TEST"/>
-    <property name="path.eslint.executable" value="${path.root}/node_modules/.bin/eslint"/>
+    <property name="path.eslint.executable" value="${path.node_modules.bin}/eslint"/>
     <property name="path.framework" value="${project.basedir}"/>
+    <property name="path.framework.assets" value="${project.basedir}./assets"/>
```
[the remark this answers is not in the document]
Comment (pesektomas) [reply packages/framework/build.xml]: Ou yes, thank you 17fb0a8ce8fa9e3f49c28df22d6a895c40532945

File: packages/framework/build.xml [hunk 8]
```diff
@@ -686,12 +689,29 @@
         </exec>
     </target>
 
+    <target name="npm-build" description="Create production build.">
```
[the remark this answers is not in the document]
Comment (pesektomas) [reply packages/framework/build.xml]: e5792229eebb36f2019d8d75ce2faae921b7b4e0

File: packages/framework/build.xml [hunk 9]
```diff
@@ -686,12 +689,29 @@
...
+        </exec>
+    </target>
+
     <target name="npm-export-translations" description="Exported used translations for javascripts." hidden="true">
         <exec executable="${path.npm.executable}" dir="${path.root}" logoutput="true" passthru="true" checkreturn="true">
             <arg value="run"/>
             <arg value="trans"/>
             <arg value="--"/>
             <arg value="source-dir=${path.assets}/**/*.js"/>
+            <arg value="source-dir=${path.framework.assets}/js/*.js"/>
```
[the remark this answers is not in the document]
Comment (pesektomas) [reply packages/framework/build.xml]: Grr, thanks 554676e1c74f749d51b075cf52dd64d3e067ea9c

File: packages/framework/assets/js/admin/components/CategoryDeleteConfirm.js [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+import Window from '../utils/Window';
+import Translator from 'bazinga-translator';
+
+export default class CategoryDeleteConfirm {
+
+    static init ($container) {
+        $container.filterAllNodes('.js-category-delete-confirm').click((event) => {
+            // eslint-disable-next-line no-new
+            new Window({
+                content: Translator.trans('Do you really want to delete this category?'),
```
[the remark this answers is not in the document]
Comment (pesektomas) [reply packages/framework/assets/js/admin/components/CategoryDeleteConfirm.js:11]: I so sorry, I broke rebase - e68b13b88a9abc1d30a52ca8965850125b01f045

File: packages/framework/build.xml [hunk 11]
```diff
@@ -686,12 +689,29 @@
...
     </target>
 
+    <target name="npm-build" description="Create production build.">
+        <exec executable="${path.npm.executable}" dir="${path.root}" logoutput="true" passthru="true" checkreturn="true">
+            <arg value="run"/>
+            <arg value="build"/>
+        </exec>
+    </target>
+
+    <target name="npm-dev" description="Compile assets once.">
```
[the remark this answers is not in the document]
Comment (pesektomas) [reply packages/framework/build.xml]: I so sorry, I broke rbase - e68b13b88a9abc1d30a52ca8965850125b01f045

File: project-base/assets/js/commands/translations/parseFile.js [hunk 12]
```diff
@@ -40,7 +40,7 @@ function parseFile (filePath) {
         });
         return translations;
     } catch (e) {
-        console.log(e);
+        throw e;
```
[the remark this answers is not in the document]
Comment (pesektomas) [reply project-base/assets/js/commands/translations/parseFile.js]: omg f660b2f1c2978c88b3e959044f361c49ebd7b21d

File: project-base/assets/js/commands/translations/process.js [hunk 13]
```diff
@@ -11,10 +11,6 @@ function process (dirWithJsFiles, dirWithTranslations, outputDirForExportedTrans
         const translations = [];
 
         filePaths.map(filePath => {
-            if (filePath.match(/(\w*)\.js$/) === null) {
-                return;
-            }
-
```
[the remark this answers is not in the document]
Comment (pesektomas) [reply project-base/assets/js/commands/translations/process.js:17]: Yes, this function has parameter `dirWithJsFiles` and this parameter is glob pattern `../**/*.js`

File: project-base/assets/js/frontend/components/NewsletterSubscriptionForm.js [hunk 14]
```diff
@@ -32,6 +44,21 @@ export default class NewsletterSubscriptionForm {
         }
     }
 
+    static formatErrors (arrayOfErrors) {
```
[the remark this answers is not in the document]
Comment (pesektomas) [reply project-base/assets/js/frontend/components/NewsletterSubscriptionForm.js:47]: I don't think so. The `CustomizeBundle.getFormattedFormErrors` formats error from element and return errors as another element.

File: project-base/assets/js/frontend/components/NewsletterSubscriptionForm.js [hunk 15]
```diff
@@ -2,13 +2,19 @@ import Ajax from 'framework/common/utils/Ajax';
...
+import CustomizeBundle from 'framework/common/validation/customizeBundle';
 
 const subscriptionFormSelector = 'form[name="subscription_form"]';
 
 export default class NewsletterSubscriptionForm {
 
     ajaxSubmit (event, newsletterSubscriptionForm) {
         event.preventDefault();
+
+        if (CustomizeBundle.isFormValid($('form[name="subscription_form"]')) === false) {
```
[the remark this answers is not in the document]
Comment (pesektomas) [reply project-base/assets/js/frontend/components/NewsletterSubscriptionForm.js:14]: This is a very strange use case. We want to send form by ajax, so we add the submit event to the form (the second, because jsFormValidator adds the first event). So JsFormValidator validates the form and displays error window (event 1), followed by our event 2, which sends an ajax request (incorrectly because the form is not valid) and displays next error window. So I will submit ajax only after I check that the form is valid.

File: project-base/assets/js/frontend/components/Repeater.js [hunk 16]
```diff
@@ -8,9 +8,9 @@ export default class Repeater {
     startAutorepeat ($input, eventString) {
```
[the remark this answers is not in the document]
Comment (pesektomas) [reply project-base/assets/js/frontend/components/Repeater.js:8]: The variable with dollar should be a element.

```
const firstNameElementName = 'input[name="firstName"]';
const $firstNameElement = $(firstNameElementName);
```

File: packages/framework/build.xml (hunk 7 above)
[the remark this answers is not in the document]
Comment (pesektomas) [reply packages/framework/build.xml]: :+1: 17fb0a8ce8fa9e3f49c28df22d6a895c40532945

File: project-base/webpack.config.js (hunk 3 above)
Comment (pesektomas) [reply project-base/webpack.config.js:10]: This is broken rebase e68b13b88a9abc1d30a52ca8965850125b01f045

File: project-base/src/Controller/Front/NewsletterController.php (hunk 6 above)
Comment (pesektomas) [reply project-base/src/Controller/Front/NewsletterController.php:79]: 288caf2bc38388cab13abf12283b6df44f5f6a2d

File: packages/framework/src/Model/Newsletter/NewsletterFacade.php [hunk 17]
```diff
@@ -48,6 +49,8 @@ public function addSubscribedEmail($email, $domainId)
             $newsletterSubscriber = $this->newsletterSubscriberFactory->create($email, new DateTimeImmutable(), $domainId);
             $this->em->persist($newsletterSubscriber);
             $this->em->flush($newsletterSubscriber);
+        } else {
+            throw new EmailWasAlreadySubscribedException();
```
[the remark this answers is not in the document]
Comment (pesektomas) [reply packages/framework/src/Model/Newsletter/NewsletterFacade.php]: 5a0a8da707e10b9fd774e7a0ec28d038ec6383bd

File: project-base/src/Controller/Front/NewsletterController.php (hunk 1 above)
Comment (pesektomas) [reply project-base/src/Controller/Front/NewsletterController.php]: 5a0a8da707e10b9fd774e7a0ec28d038ec6383bd

File: project-base/src/Controller/Front/NewsletterController.php (hunk 2 above)
Comment (pesektomas) [reply project-base/src/Controller/Front/NewsletterController.php]: this is unnecessary, it was an attempt - 27223429b561378717bd3b0702a3ff030822cb8d

File: project-base/src/Controller/Front/NewsletterController.php (hunk 5 above)
Comment (pesektomas) [reply project-base/src/Controller/Front/NewsletterController.php:72]: 27223429b561378717bd3b0702a3ff030822cb8d

File: project-base/assets/js/frontend/components/NewsletterSubscriptionForm.js (hunk 15 above)
[the remark this answers is not in the document]
Comment (grossmannmartin) [reply project-base/assets/js/frontend/components/NewsletterSubscriptionForm.js:14]: I understand, thanks for the clarification.

File: project-base/assets/js/frontend/components/Repeater.js (hunk 16 above)
[the remark this answers is not in the document]
Comment (grossmannmartin) [reply project-base/assets/js/frontend/components/Repeater.js:8]: thank you, now I get it :)

File: project-base/translations/messages.en.po [hunk 18]
```diff
@@ -175,6 +175,9 @@ msgstr ""
 msgid "Email"
 msgstr ""
 
+msgid "Email was already subscribed"
+msgstr ""
```
Comment (grossmannmartin) [inline project-base/translations/messages.en.po:179]: this is no longer used ;)

Comment (pesektomas) [reply project-base/translations/messages.en.po]: 3aa7b9914e9355409269d5cd536a1244a17c1f31
